merkletree.go

```diff
@@ -0,0 +1,274 @@
+	}
+	return nil, nil, ErrInvalidTree
+}
+
+func (m *MerkleTree) Set(key string, value []byte) error {
+	index := computePrivateIndex(key)
+	toAdd := userLeafNode{
+		key:        key,
+		value:      value,
+		index:      index,
+		commitment: crypto.Digest(m.salt, []byte(key), value),
+	}
+
+	return m.insertNode(index, &toAdd)
+}
+
+// Private Index calculation function
+// would be replaced with Ismail's VRF implementation
+func computePrivateIndex(key string) []byte {
+	return crypto.Digest([]byte(key))
+}
+
+func (m *MerkleTree) insertNode(key []byte, toAdd *userLeafNode) error {
+	position := 0
+	var nodePointer interface{}
+	nodePointer = m.root
+
+insertLoop:
+	for {
+		toAdd.level++
+		switch nodePointer.(type) {
+		case *userLeafNode:
+			// reached a "bottom" of the tree.
+			// add a new interior node and push the previous leaf down
+			// then continue insertion
+			currentNodeUL := nodePointer.(*userLeafNode)
+			if currentNodeUL.parent == nil {
+				return ErrInvalidTree
+			}
+
+			leftBranch := &emptyNode{
+				node: node{
+					level: currentNodeUL.level + 1,
+				},
+			}
+			rightBranch := &emptyNode{
+				node: node{
+					level: currentNodeUL.level + 1,
+				},
+			}
+			newInteriorNode := interiorNode{
+				node: node{
+					parent: currentNodeUL.parent,
+					level:  currentNodeUL.level,
+				},
+				leftChild:  leftBranch,
+				rightChild: rightBranch,
+			}
+			leftBranch.parent = &newInteriorNode
+			rightBranch.parent = &newInteriorNode
+
+			if currentNodeUL.key == toAdd.key {
+				// replace the value
+				currentNodeUL.value = toAdd.value
+				return nil
+			}
+
+			currentNodeKey := computePrivateIndex(currentNodeUL.key)
+
+			currentNodeUL.index = currentNodeKey
+
+			direction := util.GetNthBit(currentNodeKey, position)
+			if direction {
+				newInteriorNode.rightChild = currentNodeUL
+			} else {
+				newInteriorNode.leftChild = currentNodeUL
+			}
+			currentNodeUL.level++
+			currentNodeUL.parent = &newInteriorNode
+			if newInteriorNode.parent.(*interiorNode).leftChild == nodePointer {
+				newInteriorNode.parent.(*interiorNode).leftChild = &newInteriorNode
+			} else {
+				newInteriorNode.parent.(*interiorNode).rightChild = &newInteriorNode
+			}
+			nodePointer = &newInteriorNode
+			toAdd.level--
+		case *interiorNode:
+			currentNodeI := nodePointer.(*interiorNode)
+			direction := util.GetNthBit(key, position)
+
+			if direction { // go right
+				currentNodeI.rightHash = nil
+				if currentNodeI.rightChild.isEmpty() {
+					currentNodeI.rightChild = toAdd
+					toAdd.parent = currentNodeI
+					break insertLoop
+				} else {
+					nodePointer = currentNodeI.rightChild
+				}
+			} else { // go left
+				currentNodeI.leftHash = nil
+				if currentNodeI.leftChild.isEmpty() {
+					currentNodeI.leftChild = toAdd
+					toAdd.parent = currentNodeI
+					break insertLoop
+				} else {
+					nodePointer = currentNodeI.leftChild
+				}
+			}
+			position++
+		default:
+			return ErrInvalidTree
```

Comment:
Should probably just panic here and stop returning errors altogether in this method.